code/xcvm/cosmwasm/contracts/gateway/src/contract/ibc/ics20.rs

```diff
@@ -35,64 +39,130 @@ pub(crate) fn handle_bridge_forward(
-		.add_attribute("to_network_id", msg.to.to_string())
-		.add_attribute(
-			"assets",
-			serde_json_wasm::to_string(&packet.assets)
-				.map_err(|_| ContractError::FailedToSerialize)?,
-		)
-		.add_attribute(
-			"program",
-			serde_json_wasm::to_string(&packet.program)
-				.map_err(|_| ContractError::FailedToSerialize)?,
-		);
-	if !packet.salt.is_empty() {
-		let salt_attr = Binary::from(packet.salt.as_slice()).to_string();
-		event = event.add_attribute("salt", salt_attr);
-	}
-
-	let coin = Coin::new(amount.0, route.local_native_denom.clone());
+	let (msg, event) = if let Ok(transfer_shortcut) =
+		ibc_ics_20_transfer_shortcut(deps.as_ref(), &msg)
+	{
+		let mut _event = make_event("bridge")
+			.add_attribute("to_network_id", msg.to.to_string())
+			.add_attribute("shortcut", "ics20-transfer");
+
+		unimplemented!("add tracking lock for funds return usual cosmos message to transfer as defined in {:?}", transfer_shortcut);
+	} else {
+		let route: IbcIcs20Route = get_this_route(deps.storage, msg.to, *local_asset)?;
+		state::tracking::bridge_lock(deps.storage, (msg.clone(), route.clone()))?;
+
+		let asset = msg
+			.msg
+			.assets
+			.0
+			.get(0)
+			.map(|(_, amount)| (route.on_remote_asset, *amount))
+			.expect("not empty");
+
+		let packet = XcPacket {
+			interpreter: String::from(info.sender).into_bytes(),
+			user_origin: msg.interpreter_origin.user_origin.clone(),
+			salt: msg.msg.salt,
+			program: msg.msg.program,
+			assets: vec![asset].into(),
+		};
+
+		deps.api.debug(&format!(
+			"cvm::gateway::ibc::ics20 route {}",
+			&serde_json_wasm::to_string(&route)?
+		));
+
+		let mut event = make_event("bridge")
+			.add_attribute("to_network_id", msg.to.to_string())
+			.add_attribute(
+				"assets",
+				serde_json_wasm::to_string(&packet.assets)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			)
+			.add_attribute(
+				"program",
+				serde_json_wasm::to_string(&packet.program)
+					.map_err(|_| ContractError::FailedToSerialize)?,
+			);
+
+		if !packet.salt.is_empty() {
+			let salt_attr = Binary::from(packet.salt.as_slice()).to_string();
+			event = event.add_attribute("salt", salt_attr);
+		}
 
-	let msg = to_cw_message(deps.as_ref(), deps.api, coin, route, packet, block)?;
+		let coin = Coin::new(amount.0, route.local_native_denom.clone());
+
+		match route.gateway_to_send_to.clone() {
+			GatewayId::CosmWasm { contract, .. } => {
+				let msg = to_cosmwasm_message(
+					deps.as_ref(),
+					deps.api,
+					coin,
+					route,
+					packet,
+					block,
+					contract,
+				)?;
+				(msg, event)
+			},
+			GatewayId::Evm { .. } => Err(ContractError::NotImplemented)?,
+		}
+	};
 
 	Ok(Response::default()
 		.add_event(event)
 		.add_submessage(SubMsg::reply_on_success(msg, ReplyId::TransportSent.into())))
 }
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[serde(rename_all = "snake_case")]
+pub struct Shortcut {
```

Comment:
can  you please add a comment what shortcut is for?

Reply:
Will add. In Short, shortcut is `instead of sending full program to target network send just simple remote execution request. for example many platforms support transfer and exchange, so it is cheaper to send just native instuction than full program.`